File: compare/compare_MATLABv1_21.m
%% Set up input conditions
PARvalues = [2250 2100 8.1 400 405 200];
PARTYPEs = 1:6;
pHSCALEIN_opts = 1:4;
K1K2CONSTANTS_opts = 1:14;
KSO4CONSTANTS_opts = 1:4;
KFCONSTANT_opts = 1:2;
SALvalue = 33.1;
[P1, P2, P1type, P2type, sal, pHscales, K1K2, KSO4_only, KSO4, KF, ...
    BSal] = CO2SYSigen(PARvalues, PARTYPEs, SALvalue, pHSCALEIN_opts, ...
    K1K2CONSTANTS_opts, KSO4CONSTANTS_opts, KFCONSTANT_opts);
tempin = 24;
tempout = 12;
presin = 0;
presout = 1647;
si = 50;
phos = 10;
nh3 = 15;
h2s = 5;

% Run CO2SYS
xrow = 1 + 210; % just do one row, or...
% xrow = 1:numel(P1); % ... do all rows (do this for saving output file)
P1 = P1(xrow);
P2 = P2(xrow);
P1type = P1type(xrow);
P2type = P2type(xrow);
sal = sal(xrow);
pHscales = pHscales(xrow);
K1K2 = K1K2(xrow);
KSO4 = KSO4(xrow);
KF = KF(xrow);
BSal = BSal(xrow);
tic
[DATA, HEADERS] = ...
    CO2SYSv1_21(P1, P2, P1type, P2type, sal, tempin, tempout, presin, ...
    presout, si, phos, nh3, h2s, pHscales, K1K2, KSO4, KF, BSal);
toc

%% Extract and save outputs
clear co2s
for V = 1:numel(HEADERS)
    co2s.(HEADERS{V}) = DATA(:, V);
end % for V
co2s.PAR1 = P1;
co2s.PAR2 = P2;
co2s.PAR1TYPE = P1type;
co2s.PAR2TYPE = P2type;
co2s.KSO4CONSTANTS = KSO4_only;
##save('data/MATLAB_CO2SYSv1_21.mat', 'co2s')

File: compare/compare_MATLABv1_21.py
import numpy as np
from scipy.io import loadmat
from PyCO2SYS import CO2SYS
from time import time

# Import input conditions: "MATLAB_CO2SYSv1_21.mat" was generated in MATLAB
# using the script "compare_MATLABv1_21.m".
matfile = loadmat('compare/data/MATLAB_CO2SYSv1_21.mat')['co2s']
(P1, P2, P1type, P2type, sal, tempin, tempout, presin, presout, phos, si,
    pHscales, K1K2, KSO4, KF, nh3, h2s) = [matfile[var][0][0] for var in
    ['PAR1', 'PAR2', 'PAR1TYPE', 'PAR2TYPE', 'SAL', 'TEMPIN', 'TEMPOUT',
     'PRESIN', 'PRESOUT','PO4', 'SI', 'pHSCALEIN', 'K1K2CONSTANTS',
     'KSO4CONSTANTS', 'KFCONSTANT', 'NH3', 'H2S']]
co2inputs = [P1, P2, P1type, P2type, sal, tempin, tempout, presin, presout,
             si, phos, pHscales, K1K2, KSO4]
# xrow = 210 # just do one row, or...
xrow = range(len(P1)) # ... do all rows
co2inputs = [inp[xrow] for inp in co2inputs]

# Run CO2SYS in Python
go = time()
co2py = CO2SYS(*co2inputs, NH3=nh3[xrow], H2S=h2s[xrow], KFCONSTANT=KF[xrow])
print('PyCO2SYS runtime = {:.6f} s'.format(time() - go))
# Extract dict output if PyCO2SYS.original was used
if np.shape(co2py) == (4,):
    co2py = co2py[0]

# Prepare MATLAB results to compare
pyvars = ['NH3Alkin', 'NH3Alkout', 'H2SAlkin', 'H2SAlkout', 'KSO4CONSTANT',
          'KFCONSTANT', 'BORON', 'NH3', 'H2S', 'KNH3input', 'KNH3output',
          'KH2Sinput', 'KH2Soutput', 'gammaTCin', 'betaTCin', 'omegaTCin',
          'gammaTAin', 'betaTAin', 'omegaTAin', 'isoQin', 'isoQapprox_in',
          'gammaTCout', 'betaTCout', 'omegaTCout', 'isoQout', 'psi_in',
          'gammaTAout', 'betaTAout', 'omegaTAout', 'isoQapprox_out', 'psi_out']
co2mat = {var: matfile[var][0][0].ravel()[xrow] for var in co2py.keys()
          if var not in pyvars}
# Differences between PyCO2SYS and MATLAB v1.21
co2diff = {var: co2py[var] - co2mat[var] for var in co2mat.keys()}
co2maxdiff = {var: np.max(np.abs(co2diff[var])) for var in co2mat.keys()}
